Rebased 9 bookit commits (WIP)

dm opened the pull request:
### Likely won't merge this, but wanted to distill and comment on the changes to be brought into the develop branch.


Initial changes for bookit-specific build.
Bunch of related changes I made while trying to get ECS to build/run bookit-server, taking into account the SSM parameter store as a source for secrets.
- Appropriate task definition.  Note that it runs special run-in-aws.sh bookit-server start script.
- IAM roles assigned
- Get Task/Service/ELB ports assigned properly (or at least working)
- A number of configuration changes to allow more flexibility, which was needed during my many test runs.

Pass in repository stuff via top-level parameters.
Keep bucket names unique.  Using region in the name seems a little goofy, but it's an easy way to get the uniqueness.
Hack to prevent issue that happened due to a non-atomic failed deletion in Cfn that resulted in a hung stack that couldn't be deleted due to a deleted role.
Don't try to delete the repo associated with the stack -- it will probably have images in it and will fail deletion.
Stop using deployment-pipeline.yml (don't want to use CodePipeline/CodeBuild at this point).  Thus, pull repo "local", and create Service stack directly (with desired count of '0' to start).
Add simplest-possible DNS record creation.
Put ECS instances in private subnet.
Surface ECS cluster size.
Rename bucket prefix to awsrig
Expect S3 bucket creation to fail
Avoid seeing error on multiple executions

dm commented on Makefile:
```diff
@@ -77,16 +77,22 @@ create-foundation: upload
 			"ParameterKey=AppStackName,ParameterValue=${PROJECT}-${ENV}-${NAME_SUFFIX}-app-${APP}" \
-			"ParameterKey=BuildArtifactsBucket,ParameterValue=awsrig.${PROJECT}.${NAME_SUFFIX}.build" \
+			"ParameterKey=BuildArtifactsBucket,ParameterValue=awsrig.${PROJECT}.${NAME_SUFFIX}.${REGION}.build" \
+			"ParameterKey=EcsInstanceType,ParameterValue=t2.small" \
```
Should be using some ENV var - we can have some acceptable sane defaults though.

dm commented on Makefile:
```diff
@@ -77,16 +77,22 @@ create-foundation: upload
 			"ParameterKey=Region,ParameterValue=${REGION}" \
+			"ParameterKey=RepositoryAuthToken,ParameterValue=${REPO_TOKEN}" \
+			"ParameterKey=RepositoryBranch,ParameterValue=${REPO_BRANCH}" \
+			"ParameterKey=RepositoryName,ParameterValue=${REPO}" \
```
I had added and then removed using Github, but likely can bring it back. TBC

mathomas replied:
And we have pulled it out for now, since Travis runs the build in bookit's case.

dm commented on aws/app/deployment-pipeline.yaml:
```diff
@@ -29,18 +29,29 @@ Parameters:
 Resources:
   Repository:
     Type: AWS::ECR::Repository
+    DeletionPolicy: Retain
```
Alternatively, we can add an AWS CLI check that there are no images: `aws ecr list-images --repository-name <repo-name>`

mathomas replied:
That would be nice.  Of course, in our divergent version of this repo, the ECR management is now a script (I don't necessarily agree with that, but that's how it is for now).

dm commented on aws/app/deployment-pipeline.yaml:
```diff
@@ -29,18 +29,29 @@ Parameters:
     Type: AWS::IAM::Role
+    # Hate to do this, but if stack deletion fails after this role has been deleted
+    # then you'll never be able to complete deletion of the service stack.
+    DeletionPolicy: Retain
```
I'm sure there's a way to ensure the CFnExRole isn't deleted until the very end.
I'll try and investigate this!

mathomas replied:
Would be cool.  It's a real bitch when the role is deleted out from other its dependencies.  Amazon has a big bug here, I think.

dm replied:
From what I remember, it's likely as simple as adding `DependsOn: CloudFormationExecutionRole` on the `Pipeline` resource; will add that to try on

dm commented on aws/app/deployment-pipeline.yaml:
```diff
@@ -103,6 +114,11 @@ Resources:
                   - ecr:InitiateLayerUpload
                   - ecr:UploadLayerPart
                   - ecr:CompleteLayerUpload
+              # TODO:  this needs to be tightened-up.
```
👍

dm commented on aws/app/ecs-cluster.yaml:
```diff
@@ -96,8 +99,8 @@ Resources:
       VPCZoneIdentifier:
-        - Fn::ImportValue: !Sub "${FoundationStackName}--SubnetPublicA"
-        - Fn::ImportValue: !Sub "${FoundationStackName}--SubnetPublicB"
+        - Fn::ImportValue: !Sub "${FoundationStackName}--SubnetPrivateA"
```
👍

dm commented on aws/app/load-balancer.yaml:
```diff
@@ -51,7 +59,7 @@ Resources:
       VpcId:
         Fn::ImportValue: !Sub "${FoundationStackName}--VpcId"
-      Port: 80
+      Port: 8888
```
👎 

Listening port depends on the app; we should move it to a config setting rather than hardcode it.

mathomas replied:
Hah.  Yes.  All good intentions.  Didn't expect you to accept every commit!  We have a card for this on our project board.

dm commented on aws/app/load-balancer.yaml:
```diff
@@ -77,6 +85,17 @@ Resources:
         - TargetGroupArn: !Ref TargetGroup
           Type: forward
 
+  elbDNS:
```
Was going to add an entries for all apps, but this is fine ATM 👍

dm commented on aws/app/main.yaml:
```diff
@@ -37,21 +41,45 @@ Parameters:
     Description: Region
     Type: String
 
+  EcsInstanceType:
```
Should add `AllowedValues` for this

mathomas replied:
Agreed, sort-of.  I decided it wasn't my job to decide what was allowed, per se.  On the other hand, is there a pseudo-parameter for it?  I didn't get around to checking yet.

dm replied:
I would suggest that you limit to acceptable types to ensure that the [wrong - or too expensive - instances](https://aws.amazon.com/ec2/instance-types/) aren't used. I.e.: `t2.micro, t2.small, t2.medium, m3.medium, m3.large, m3.xlarge, m3.2xlarge`

dm commented on aws/app/main.yaml:
```diff
@@ -37,21 +41,45 @@ Parameters:
     Type: String
 
 Resources:
-  CodeCommitRepo:
```
The CodeCommit will likely be kept; even if not used for a particular App. Again Ansible version will be better to dynamically enable/disable in the template :)

mathomas replied:
Yeah, we're starting to get into bookit-specific stuff here.

dm commented on aws/app/main.yaml:
```diff
@@ -60,6 +88,8 @@ Resources:
       Parameters:
         AppStackName: !Ref AppStackName
         FoundationStackName: !Ref FoundationStackName
+        DnsHostedZoneName: riglet.io
```
Export `DnsHostedZoneName` and `DnsSubdomainName ` to variables

mathomas replied:
Yep.

dm commented on aws/app/main.yaml:
```diff
@@ -71,34 +101,49 @@ Resources:
+#        RepositoryAuthToken: !Ref RepositoryAuthToken
+#        TargetGroup: !GetAtt LoadBalancer.Outputs.TargetGroup
+
+  Repository:
```
Duplicate entry for `AWS::ECR::Repository`?

dm commented on aws/app/service.yaml:
```diff
@@ -46,40 +46,52 @@ Resources:
       LoadBalancers:
         - ContainerName: app-deployment
-          ContainerPort: 80
+          ContainerPort: 8888
```
Use from variable

mathomas replied:
Yep.

dm commented on aws/app/service.yaml:
```diff
@@ -46,40 +46,52 @@ Resources:
+            Version: 2012-10-17
+            Statement:
+              # TODO:  this needs to be tightened-up.
+              - Resource: !Sub arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/*
```
At least define a prefix here (after `parameter/*`) for at least `ProjectName`?

dm reviewed (commented):
Some thoughts

dm committed:
Rebased 9 bookit commits

Initial changes for bookit-specific build.
Bunch of related changes I made while trying to get ECS to build/run bookit-server, taking into account the SSM parameter store as a source for secrets.
- Appropriate task definition.  Note that it runs special run-in-aws.sh bookit-server start script.
- IAM roles assigned
- Get Task/Service/ELB ports assigned properly (or at least working)
- A number of configuration changes to allow more flexibility, which was needed during my many test runs.
Pass in repository stuff via top-level parameters.
Keep bucket names unique.  Using region in the name seems a little goofy, but it's an easy way to get the uniqueness.
Hack to prevent issue that happened due to a non-atomic failed deletion in Cfn that resulted in a hung stack that couldn't be deleted due to a deleted role.
Don't try to delete the repo associated with the stack -- it will probably have images in it and will fail deletion.
Stop using deployment-pipeline.yml (don't want to use CodePipeline/CodeBuild at this point).  Thus, pull repo "local", and create Service stack directly (with desired count of '0' to start).
Add simplest-possible DNS record creation.
Put ECS instances in private subnet.
Surface ECS cluster size.
Rename bucket prefix to awsrig
Expect S3 bucket creation to fail
Avoid seeing error on multiple executions

dm committed:
Change LB port to 443

Remove some app defaults
Change service container port to use a Parameter

dm committed:
WIP: Bookit, need review, likely too specific
